storagenode/nodestats/cache.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package nodestats
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"storj.io/storj/internal/date"
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/storagenode/reputation"
+	"storj.io/storj/storagenode/storageusage"
+	"storj.io/storj/storagenode/trust"
+)
+
+var (
+	// NodeStatsCacheErr defines node stats cache loop error
+	NodeStatsCacheErr = errs.Class("node stats cache error")
+)
+
+// Cache runs cache loop and stores reputation stats
+// and storage usage into db
+type Cache struct {
+	log *zap.Logger
+
+	service        *Service
+	trust          *trust.Pool
+	reputationDB   reputation.DB
+	storageusageDB storageusage.DB
+
+	reputationCycle *sync2.Cycle
+	storageCycle    *sync2.Cycle
+}
+
+// NewCache creates new caching service instance
+func NewCache(log *zap.Logger, service *Service, trust *trust.Pool, reputationDB reputation.DB, storageusageDB storageusage.DB) *Cache {
+	return &Cache{
+		log:             log,
+		service:         service,
+		trust:           trust,
+		reputationDB:    reputationDB,
+		storageusageDB:  storageusageDB,
+		reputationCycle: sync2.NewCycle(time.Hour * 4),
+		storageCycle:    sync2.NewCycle(time.Hour * 12),
+	}
+}
+
+// Run runs loop
+func (cache *Cache) Run(ctx context.Context) error {
+	var group errgroup.Group
+
+	cache.reputationCycle.Start(ctx, &group, func(ctx context.Context) error {
+		err := cache.CacheReputationStats(ctx)
+		if err != nil {
+			cache.log.Error("Get stats query failed", zap.Error(err))
+		}
+
+		return nil
+	})
+	cache.storageCycle.Start(ctx, &group, func(ctx context.Context) error {
```

Comment:
why use `cycle.Start()` instead of `cycle.Run()`?

Reply:
Because `cycle.Start` has `errgroup` as input, as I have 2 cycles that should run in parallel under control of single `cache.Run` I thought that it is more appropriate to do it this way